_posts/2021-01-12-updating-foreman-inventory-with-system-facts.md

```diff
@@ -0,0 +1,323 @@
+where facts can be used. We'll illustrate the configuration on one config management system, but the same concept is applicable
+to other facts sources. Links to the related documentation will be provided.
+
+<!--more-->
+
+## Automatically update the inventory from configuration management
+
+Having to keep a big inventory up to date manually would be a lot of work. Foreman
+can receive updates from other tools used to manage the infrastructure and reflect
+them in the database. Also new Hosts can be created this way. If you use some
+configuration management system or you use a subscription manager (mostly RPM
+based distributions), you can configure it to forward data about managed hosts to Foreman. If the host does not exist in Foreman yet, it can be automatically created. If it exists, the record is updated accordingly.
+
+Typically every configuration management agent run starts with gathering facts.
+Facts are generic information about the machine, OS, networking and other areas.
+These facts can be uploaded to Foreman. They are parsed and processed, so that Foreman knows e.g.
+about the current operating system version, kernel, IP addresses, available RAM etc.
+Foreman can also update other existing objects, such as associated subnet, OS or similar, if configured to do so. Also all the raw values
+are stored for later use. We simply call them Facts.
+
+## Facts importing overview
+
+The facts upload process is similar for all fact sources. The data is collected on the Smart Proxy machine and then sent to Foreman API. The request is authenticated using an x509 certificate. For the Authorization, the Foreman verifies, it comes from a Smart Proxy host, that have related Feature enabled (e.g. Puppet, Ansible etc). Then it imports the data and stores it to the database.
+
+<p float='left' align='center'>
+  <img src="/static/images/blog_images/2021-01-12-updating-foreman-inventory-with-system-facts/facts_upload_diagram.png" width="700" />
+</p>
+
+If the data contains a new fact, the new fact is created. If there’s a fact from a previous fact import, that’s no longer present in the new set, it’s deleted. Fact values are updated if they changed in the new data payload.
+
+Once facts are imported the parsing follows. We normalize them and update Host attributes. For example, Puppet reports uptime while subscription-manager reports boot time timestamp. At this point we calculate the boot time and store that in a unified form which can be easily used for searching in Foreman inventory.
+
+We also update the host attributes, such as the operating system or even create a new OS object if it does not exist yet. As mentioned earlier, if we see facts for the host that does not exist in the Foreman database yet, Foreman can create it. This can be used to populate the inventory easily when you start with Foreman in the brownfield deployment.
```

Comment:
```suggestion
We also update the host attributes, such as the operating system or even create a new OS object if it does not exist yet. As mentioned earlier, if we see facts for the host that does not exist in the Foreman database yet, Foreman can create it. This can be used to populate the inventory easily when you start with Foreman in a brownfield deployment.
```